README.md
# SquashMe

Tired of constantly trying to find free court for your beloved squash? 
Maybe some badminton game? Fear not. Squash-Me is a package for 
automating <i>Hasta La Vista</i> - one of the biggest sport centres in Wroclaw.

With it, you can constantly fetch and reserve free courts at desired time.

## Requirements

- other packages
```
lxml==4.3.1
requests==2.21.0
```

- config.ini file
```
[USER_INFO]
# For getting page content and making reservations.
login = dummy
password = dummy

# For creating client id that is used with reservation request.
name = dummy
surname = dummy
``` 

User is prompted for information on first CLI execution 
in order to create the file at SquashMe\squasher\config.ini

The config can be later altered with --rename or manually at the given location.


## Installation
Either:
- Install directly from repository
```
pip install git+ssh://{USER}@github.com:pdamec/squash-me.git
```

- Clone repository and install with setuptools
```
Git clone {USER}@github.com:pdamec/squash-me.git
python setup.py install
```

## Usage
### Command Line

```
sq-me --court_number 4 --discipline squash --start 17:30 --end 19:00 --day 2019-03-28
```

Where:
- court_number: desired court. If not provided, SquashMe searches at all courts.
- discipline: supports squash, badminton and table_tennis. Default is squash.
- start: from what time look for available courts. Default set for 06:00 AM.
- end: until what time look for available courts. Default set for 12 AM.
- day: on what day free courts should be checked. Default is day of program execution.
- rename: Create new config.ini

IMPORTANT NOTE: start, end arguments support 24-hour format.
 
For more info, execute:

```
sq-me --help
```




CHANGELOG.md
## [1.0.0] - 2019-03-30
### Added
- SquashMe class consisting of all necessary methods to automate reservations.
- Command line
- Config file creation on first CLI execution
- Reservations for each court since 15:00 until 24:00


tests/test.py
from unittest import TestCase
from squasher.squasher import SquashMe


class TestConsole(TestCase):
    def test_main_class(self):
        SquashMe()


squasher/wrapper.py
import os
import logging
from getpass import getpass
from configparser import ConfigParser
from pprint import pformat
from .squasher import SquashMe

logger = logging.getLogger(__name__)


class SQWrapper:

    CONFIG_PATH = os.path.join(os.path.dirname(__file__), 'config.ini')

    def __init__(self, **kwargs):
        self.args = kwargs
        self.set_config()
        self.sq = SquashMe(**kwargs, config=self.get_config())

    def start_me_please(self):
        free_reservations = self.sq.show_free_reservations()

        if free_reservations:
            logger.info(pformat(free_reservations))
            user_reservations = input('Select hours to reserve (hh:mm-hh:mm or hh:mm): ')

            if self.args['court_number'] is None:
                self.set_court()
            self.reserve(user_reservations)
        else:
            logger.info('No courts available between {}-{} on {}.'.format(self.args['start'], self.args['end'],
                                                                          self.args['day']))

    def set_court(self):
        court = input('Select court (number): ')
        if court:
            self.sq.court_number = int(court)
            logger.info('Getting reservations for court: {}'.format(self.sq.court_number))
        else:
            logger.info('No court provided.')

    def reserve(self, user_reservations):
        if user_reservations:
            user_reservations = list(self.sq.get_user_reservations(user_reservations))
            print(user_reservations)
            reservation_slots = self.sq.create_reservation_payload(user_reservations)
            self.sq.request_reservations(reservation_slots)
        else:
            logger.info('No user reservations provided.')

    def set_config(self):
        if not os.path.isfile(SQWrapper.CONFIG_PATH) or self.args['is_rename']:
            login = input('Provide login to Hasta La Vista: ')
            password = getpass()
            username = input('Provide name and surname (Name Surname) '
                             'that were used to create Hasta La Vista account: ')
            self.create_config(login, password, username)

    @staticmethod
    def create_config(login, password, username):
        username = username.lower().strip().split(' ')

        config = ConfigParser()
        config['USER_INFO'] = dict(login=login, password=password, name=username[0], surname=username[1])
        with open(SQWrapper.CONFIG_PATH, 'w') as configfile:
            config.write(configfile)

    @staticmethod
    def get_config():
        config = ConfigParser()
        config.read(SQWrapper.CONFIG_PATH)

        name = config['USER_INFO']['NAME']
        surname = config['USER_INFO']['SURNAME']
        user_id = '{}{}'.format(surname, name[0]).lower()
        return dict(login=config['USER_INFO']['LOGIN'], password=config['USER_INFO']['PASSWORD'], user_id=user_id)


squasher/squasher.py
import re
from datetime import datetime, date, timedelta
import logging
from requests import Session
from lxml import html
from .settings import login_page, admin_page


logger = logging.getLogger(__name__)


class SquashMe:

    def __init__(self, court_number=None, discipline='squash', start="6:00", end="23:30",
                 day=datetime.today().strftime('%Y-%m-%d'), **kwargs):

        if 'config' in kwargs:
            self.config = kwargs['config']

        self.session = None
        self.court_number = court_number
        self.discipline = discipline
        self.start = self.convert_time(start)
        self.end = self.convert_time(end)
        self.day = day
        self.parser = self.create_parser()

    @property
    def free_reservations(self):
        return self.get_free_reservations()

    def create_parser(self):
        payload = dict(log=self.config['login'], pwd=self.config['password'])

        self.session = Session()
        self.session.post(login_page, data=payload)
        content = self.load_reservation_table()
        return html.fromstring(content)

    def get_costs(self):
        regex = r'\d{2} - \d{2}|\d{2} \w{2}'
        table_headers = [i.strip() for i in self.parser.getchildren()[0].itertext() if re.match(regex, i)]
        return {hours: cost for hours, cost in zip(table_headers[0::2], table_headers[1::2])}

    def get_free_reservations(self):
        free_reservations = self.parser.find_class('rez rez_wolne')
        for i in free_reservations:
            court = i.getparent().getchildren()[0].text
            for a in i:
                free_since = self.convert_time(a.attrib.get('data-godz_od'))
                free_until = self.convert_time(a.attrib.get('data-godz_do'))
                if free_since >= self.start and free_until <= self.end:
                    yield {'free_since': a.attrib.get('data-godz_od'), 'free_until': a.attrib.get('data-godz_do'),
                           'id': i.getparent().attrib['data-obie_id'], 'court': court}

    @staticmethod
    def convert_time(res_time):
        # WA for comparing 00:00 hour in get_free_reservations().
        now = date.today()
        if res_time == "00:00":
            now = now.replace(day=now.day + 1)
        return datetime.strptime('{} {}'.format(now, res_time), '%Y-%m-%d %H:%M')

    def load_reservation_table(self):
        payload = {
            'operacja': 'ShowRezerwacjeTable',
            'action': 'ShowRezerwacjeTable',
            'data': self.day,
            'obiekt_typ': self.discipline.lower(),
        }
        req = self.session.post(admin_page, data=payload)
        return req.content

    # @staticmethod
    # def _is_between(time, time_range):
    #     if time_range[1] < time_range[0]:
    #         return time >= time_range[0] or time <= time_range[1]
    #     return time_range[0] <= time <= time_range[1]
    #
    # def ttt(self):
    #     # TBD
    #     # 'godz_od': '15:00' 11:00 06:00
    #     # 'godz_do': '00:00' 20:00 15:00
    #     #
    #     self._is_between('11:00', ('06:00', '16:00'))

    def create_reservation_payload(self, user_reservations):
        print(self.court_number)
        for reservation in self.free_reservations:
            if reservation['free_since'] in user_reservations and int(reservation['court']) == self.court_number:
                yield '{}_{}_{}'.format(reservation['id'], reservation['free_since'], reservation['free_until'])

    def request_reservations(self, reservation_slots):
        for slot in reservation_slots:
            payload = dict(action='RezerwujWybraneZapisz', data=self.day)
            payload.update({'REZ[]': slot})
            self.session.post(admin_page, payload)

            parsed_payload = payload['REZ[]'].split('_')
            logger.info(f'Making reservation for {self.discipline} between {parsed_payload[1]} - {parsed_payload[2]} '
                        f'on court {self.court_number}')

        self.session.post(admin_page, dict(action='Rezerwacje4Datepicker', klie_nick=self.config['user_id']))

    def get_user_reservations(self, reservations):
        reservations = reservations.strip().split(' ')
        for reservation in reservations:
            if re.match(r'\d{2}:\d{2}-\d{2}:\d{2}', reservation):
                starting = self.convert_time(reservation.split('-')[0])
                ending = self.convert_time(reservation.split('-')[1])
                while starting < ending:
                    yield starting.strftime('%H:%M')
                    starting += timedelta(minutes=30)
            elif re.match(r'\d{2}:\d{2}', reservation):
                yield reservation
            else:
                logging.info('Wrong reservation time format: {}.'.format(reservation))

    def show_free_reservations(self):
        if self.court_number:
            reservations = []
            for reservation in self.free_reservations:
                if int(reservation['court']) == self.court_number:
                    reservations.append(reservation)
        else:
            reservations = list(self.free_reservations)
        return reservations

    def automatic_reservation(self):
        # TBD
        pass


squasher/cli.py
import argparse
import logging
from datetime import datetime
from .wrapper import SQWrapper
from squasher.settings import cli_help


def create_parser():
    today = datetime.today().strftime('%Y-%m-%d')

    parser = argparse.ArgumentParser(description='Package for sport centre automation.')
    parser.add_argument('--court_number', type=int, help=cli_help['court_number'])
    parser.add_argument('--discipline', type=str, default='squash', help=cli_help['discipline'])
    parser.add_argument('--start', type=str, default="6:00", help=cli_help['start'])
    parser.add_argument('--end', type=str, default="23:30", help=cli_help['end'])
    parser.add_argument('--day', type=str, default=today, help=cli_help['day'])
    parser.add_argument('--rename', type=bool, nargs='?', const=True, help=cli_help['rename'])
    return parser.parse_args()


def main():
    # TBD: delete.
    logging.basicConfig(level=logging.INFO)

    args = create_parser()

    wrapper = SQWrapper(court_number=args.court_number, discipline=args.discipline, start=args.start, end=args.end,
                        day=args.day, is_rename=args.rename)
    wrapper.start_me_please()
